server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java

```diff
@@ -102,316 +34,44 @@ public static void removeThreadContext(ThreadContext threadContext) {
-            doesNotNeedEncoding.set(i);
-        }
-        assert doesNotNeedEncoding.get('%') == false : doesNotNeedEncoding;
-    }
-
-    private static final Charset UTF_8 = StandardCharsets.UTF_8;
-
     /**
-     * Encode a string containing characters outside of the legal characters for an RFC 7230 quoted-string.
+     * Adds a formatted warning message as a response header on the thread context, and logs a deprecation message if the associated key has
+     * not recently been seen.
      *
-     * @param s the string to encode
-     * @return the encoded string
+     * @param key    the key used to determine if this deprecation should be logged
+     * @param msg    the message to log
+     * @param params parameters to the message
      */
-    static String encode(final String s) {
-        // first check if the string needs any encoding; this is the fast path and we want to avoid creating a string builder and copying
-        boolean encodingNeeded = false;
-        for (int i = 0; i < s.length(); i++) {
-            int current = s.charAt(i);
-            if (doesNotNeedEncoding.get(current) == false) {
-                encodingNeeded = true;
-                break;
-            }
-        }
-
-        if (encodingNeeded == false) {
-            return s;
-        }
-
-        final StringBuilder sb = new StringBuilder(s.length());
-        for (int i = 0; i < s.length();) {
-            int current = s.charAt(i);
-            /*
-             * Either the character does not need encoding or it does; when the character does not need encoding we append the character to
-             * a buffer and move to the next character and when the character does need encoding, we peel off as many characters as possible
-             * which we encode using UTF-8 until we encounter another character that does not need encoding.
-             */
-            if (doesNotNeedEncoding.get(current)) {
-                // append directly and move to the next character
-                sb.append((char) current);
-                i++;
-            } else {
-                int startIndex = i;
-                do {
-                    i++;
-                } while (i < s.length() && doesNotNeedEncoding.get(s.charAt(i)) == false);
-
-                final byte[] bytes = s.substring(startIndex, i).getBytes(UTF_8);
-                // noinspection ForLoopReplaceableByForEach
-                for (int j = 0; j < bytes.length; j++) {
-                    sb.append('%').append(hex(bytes[j] >> 4)).append(hex(bytes[j]));
-                }
-            }
-        }
-        return sb.toString();
-    }
-
-    private static char hex(int b) {
-        final char ch = Character.forDigit(b & 0xF, 16);
-        if (Character.isLetter(ch)) {
-            return Character.toUpperCase(ch);
-        } else {
-            return ch;
-        }
+    public void deprecatedAndMaybeLog(final String key, final String msg, final Object... params) {
```

Comment:
@jakelandis I am also considering renaming these as deprecatedAndMaybe is not clear enough
I think this should be just `throttleLogAndWarnOnHeader` and `logAndWarnOnHeader` 
the destination -deprecation log - is clear from the name of the field and class